### PR TITLE
[Crossgen2] Improve compilation throughput for types with many fields

### LINKED ISSUE #38259: CG2: Time out in JIT\Regression\CLR-x86-JIT\V2.0-Beta2\b399444\b399444a
Example run:
https://dev.azure.com/dnceng/public/_build/results?buildId=698082&view=ms.vss-test-web.build-test-results-tab&runId=21633748&resultId=109838&paneView=debug

Tests failing:
JIT\Regression\CLR-x86-JIT\V2.0-Beta2\b399444\b399444a
JIT\Regression\CLR-x86-JIT\V2.0-Beta2\b399444\b399444b

Platform:
Win10 x64, Linux x64

Item Log:
<pre>
Could Not Find C:\h\w\A62C0917\w\A98C097A\e\JIT\Regression\CLR-x86-JIT\V2.0-Beta2\b399444\b399444a\b399444a.dll.rsp

cmdLine:C:\h\w\A62C0917\w\A98C097A\e\JIT\Regression\CLR-x86-JIT\V2.0-Beta2\b399444\b399444a\b399444a.cmd Timed Out

Return code:      -100
Raw output file:      C:\h\w\A62C0917\w\A98C097A\e\JIT\Regression\Reports\JIT.Regression\CLR-x86-JIT\V2.0-Beta2\b399444\b399444a\b399444a.output.txt
Raw output:
BEGIN EXECUTION
        1 file(s) copied.
Response file: C:\h\w\A62C0917\w\A98C097A\e\JIT\Regression\CLR-x86-JIT\V2.0-Beta2\b399444\b399444a\\b399444a.dll.rsp
--inputbubble
C:\h\w\A62C0917\w\A98C097A\e\JIT\Regression\CLR-x86-JIT\V2.0-Beta2\b399444\b399444a\IL\b399444a.dll
-r:C:\h\w\A62C0917\w\A98C097A\e\JIT\Regression\CLR-x86-JIT\V2.0-Beta2\b399444\b399444a\IL\*.dll
-o:C:\h\w\A62C0917\w\A98C097A\e\JIT\Regression\CLR-x86-JIT\V2.0-Beta2\b399444\b399444a\\b399444a.dll
--targetarch:x64
-O
-r:C:\h\w\A62C0917\p\System.*.dll
-r:C:\h\w\A62C0917\p\Microsoft.*.dll
-r:C:\h\w\A62C0917\p\mscorlib.dll
-r:C:\h\w\A62C0917\p\netstandard.dll
" "dotnet" "C:\h\w\A62C0917\p\crossgen2\crossgen2.dll" @"C:\h\w\A62C0917\w\A98C097A\e\JIT\Regression\CLR-x86-JIT\V2.0-Beta2\b399444\b399444a\\b399444a.dll.rsp" "

cmdLine:C:\h\w\A62C0917\w\A98C097A\e\JIT\Regression\CLR-x86-JIT\V2.0-Beta2\b399444\b399444a\b399444a.cmd Timed Out
Test Harness Exitcode is : -100
To run the test:
> set CORE_ROOT=C:\h\w\A62C0917\p
> C:\h\w\A62C0917\w\A98C097A\e\JIT\Regression\CLR-x86-JIT\V2.0-Beta2\b399444\b399444a\b399444a.cmd

</pre>

### PR DESCRIPTION
**Pull request by nattress**
`IsLayoutFixedInCurrentVersionBubble` is a fairly expensive algorithm and can be called very frequently for the same type during JIT compilation. Store the computed result so future lookups are fast.

Fixes https://github.com/dotnet/runtime/issues/38259

cc @dotnet/crossgen-contrib

### CODE REVIEW

**Review by trylek — APPROVED** [review]
Looks beautiful, thanks Simon for fixing this so quickly!

File: src/coreclr/src/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs [hunk 1]
```diff
@@ -423,6 +429,9 @@ public bool IsLayoutFixedInCurrentVersionBubble(TypeDesc type)
             return true;
         }
 
+        public bool IsLayoutFixedInCurrentVersionBubble(TypeDesc type) =>
+            _computedFixedLayoutTypes.GetOrAdd(type, (_) => IsLayoutFixedInCurrentVersionBubbleInternal(type));
```
Comment (MichalStrehovsky) [inline src/coreclr/src/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs:433]: This avoids allocating the capture - we still allocate the delegate (and don't really have to), but it's a nice 50% reduction:

```suggestion
            _computedFixedLayoutTypes.GetOrAdd(type, (t) => IsLayoutFixedInCurrentVersionBubbleInternal(t));
```

Comment (am11) [reply src/coreclr/src/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs:433]: Also, `GetOrAdd` [is non-atomic](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.getoradd?view=netcore-3.1) as the write operation is not thread-safe. The delegate can execute multiple times for the same key, which can be prevented by something like this:

```c#
_computedFixedLayoutTypes.GetOrAdd(type, t =>
{
    Interlocked.Increment(ref s_someStaticInteger);
    return _computedFixedLayoutTypes.TryGetValue(t, out bool isLayoutFixed) ?
        isLayoutFixed : IsLayoutFixedInCurrentVersionBubbleInternal(t));
});
```

Comment (MichalStrehovsky) [reply src/coreclr/src/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs:433]: We generally don't worry about those kinds of races - the worst that can happen is that we compute the same thing twice. 95% of time the thing we're computing is actually cheap to compute anyway so chances of that mattering are pretty low.

I'm not sure the double-check would actually guarantee we don't compute the same thing twice - the delegate is called outside locks, so we could still end up computing the same things twice.

Comment (am11) [reply src/coreclr/src/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs:433]: > the worst that can happen is that we compute the same thing twice.

Typically, it is proportional to degree of parallelism.
>The delegate is called outside locks, so we could still end up computing the same things twice.

The lock before second check will prevent it from happening.

Comment (nattress) [reply src/coreclr/src/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs]: Thanks Michal; I removed the capture. I agree here we're not that worried about a race in this situation. The value returned by `IsLayoutFixedInCurrentVersionBubbleInternal` is going to be identical throughout a given compilation for a given `TypeDesc`, too

**Pull request merged**